File: macos/Markdown/Markdown/ChatViewModel.swift
//
//  ChatViewModel.swift
//  Markdown
//
//  The chat transcript and the one AgentClient turn in flight at a time.
//

import Foundation
import Observation

struct ChatMessage: Identifiable {
    enum Kind {
        case user(String)
        case assistant(String)
        case tool(name: String, ok: Bool?, detail: String, commit: String?)
        /// A refusal or failure, shown inline rather than as an assistant reply.
        case notice(String)
    }

    let id = UUID()
    var kind: Kind
}

@Observable
final class ChatViewModel {
    private(set) var messages: [ChatMessage] = []
    var draft = ""
    private(set) var isResponding = false
    private(set) var errorMessage: String?

    /// Recreated whenever the vault changes, so a stale session is never reused.
    private var client: AgentClient?
    private var clientVaultRoot: URL?

    /// Commit ids already reverted from this transcript, so a tool row's Undo button can't
    /// be used twice.
    private(set) var undoneCommits: Set<String> = []

    /// Separate from `client`/`clientVaultRoot`: undo is a plain vault call, not an agent turn.
    private var vault: VaultStore?
    private var vaultRoot: URL?

    func send(vaultRoot: URL) {
        let text = draft.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty, !isResponding else { return }

        guard let client = client(for: vaultRoot) else { return }

        draft = ""
        errorMessage = nil
        messages.append(ChatMessage(kind: .user(text)))
        let assistantIndex = messages.count
        messages.append(ChatMessage(kind: .assistant("")))
        isResponding = true

        let started = client.send(text) { [weak self] event in
            Task { @MainActor in
                self?.apply(event, assistantIndex: assistantIndex)
            }
        }
        if !started {
            // Can only happen if a previous turn's background thread has not finished
            // draining yet; `isResponding` normally prevents this from being reachable.
            isResponding = false
            messages.removeLast(2)
            draft = text
            errorMessage = "The assistant is still finishing the last message."
        }
    }

    /// Opens (or reuses) the session for `vaultRoot`. Fails, with a message set, when there
    /// is no API key yet.
    private func client(for vaultRoot: URL) -> AgentClient? {
        if let client, clientVaultRoot == vaultRoot {
            return client
        }
        guard let key = Keychain.apiKey(), !key.isEmpty else {
            errorMessage = "Add an Anthropic API key in Settings to use the assistant."
            return nil
        }
        guard let opened = AgentClient(vaultRoot: vaultRoot, apiKey: key) else {
            errorMessage = "Could not open the assistant for this folder."
            return nil
        }
        client = opened
        clientVaultRoot = vaultRoot
        return opened
    }

    /// Reverts an earlier tool call by the commit id shown on its row. The revert is itself a
    /// new commit (`History::revert`), so this can't lose data — it can only be called once
    /// per commit from here, tracked by `undoneCommits`, to keep the transcript honest about
    /// what's already been undone.
    func undo(commit: String, vaultRoot: URL) {
        guard !undoneCommits.contains(commit) else { return }
        guard let vault = vault(for: vaultRoot) else { return }

        do {
            try vault.undo(commit: commit)
            undoneCommits.insert(commit)
            messages.append(ChatMessage(kind: .notice("Undid that change.")))
        } catch {
            messages.append(ChatMessage(kind: .notice("Could not undo: \(error.localizedDescription)")))
        }
    }

    /// Opens (or reuses) the vault for `vaultRoot`, independent of the agent `client` above.
    private func vault(for vaultRoot: URL) -> VaultStore? {
        if let vault, self.vaultRoot == vaultRoot {
            return vault
        }
        guard let opened = try? VaultStore(root: vaultRoot) else {
            messages.append(ChatMessage(kind: .notice("Could not open the vault to undo that change.")))
            return nil
        }
        vault = opened
        self.vaultRoot = vaultRoot
        return opened
    }

    @MainActor
    private func apply(_ event: AgentEvent, assistantIndex: Int) {
        guard assistantIndex < messages.count else { return }

        switch event {
        case .text(let delta):
            if case .assistant(let existing) = messages[assistantIndex].kind {
                messages[assistantIndex].kind = .assistant(existing + delta)
            }

        case .thinking:
            // Summarised reasoning; not surfaced in this first pass.
            break

        case .toolStarted(let name):
            messages.append(ChatMessage(kind: .tool(name: name, ok: nil, detail: "Running…", commit: nil)))

        case .toolFinished(let name, let ok, let detail, let commit):
            // Matched by name from the end: tools can run more than once in a turn, so this
            // finds the most recent still-running row for that name rather than the first.
            if let index = messages.lastIndex(where: {
                if case .tool(name, nil, _, _) = $0.kind { return true }
                return false
            }) {
                messages[index].kind = .tool(name: name, ok: ok, detail: detail, commit: commit)
            }

        case .refused(let category, let explanation):
            isResponding = false
            let reason = [category, explanation].compactMap { $0 }.joined(separator: ": ")
            messages.append(ChatMessage(
                kind: .notice(reason.isEmpty ? "The assistant declined to continue." : reason)
            ))

        case .failed(let message):
            isResponding = false
            messages.append(ChatMessage(kind: .notice(message)))

        case .done:
            isResponding = false
        }
    }
}

File: win/MarkdownWin/MarkdownWin/ChatViewModel.cs
//
//  ChatViewModel.cs
//  MarkdownWin
//
//  The chat transcript and the one AgentClient turn in flight at a time. Mirrors
//  ChatViewModel.swift.
//

using System;
using System.Collections.ObjectModel;
using System.Linq;
using Microsoft.UI.Dispatching;

namespace MarkdownWin;

internal sealed class ChatViewModel : ObservableBase, IDisposable
{
    public ObservableCollection<ChatMessage> Messages { get; } = new();

    private string draft = string.Empty;
    public string Draft
    {
        get => draft;
        set => Set(ref draft, value);
    }

    private bool isResponding;
    public bool IsResponding
    {
        get => isResponding;
        private set => Set(ref isResponding, value);
    }

    private string? errorMessage;
    public string? ErrorMessage
    {
        get => errorMessage;
        private set => Set(ref errorMessage, value);
    }

    private readonly DispatcherQueue dispatcherQueue;

    /// <summary>Recreated whenever the vault changes, so a stale session is never reused.</summary>
    private AgentClient? client;
    private string? clientVaultRoot;

    public ChatViewModel(DispatcherQueue dispatcherQueue) => this.dispatcherQueue = dispatcherQueue;

    public void Send(string vaultRoot)
    {
        string text = Draft.Trim();
        if (text.Length == 0 || IsResponding)
        {
            return;
        }

        if (ClientFor(vaultRoot) is not { } current)
        {
            return; // ErrorMessage already set; no messages appended, matches Swift.
        }

        Draft = string.Empty;
        ErrorMessage = null;
        var user = ChatMessage.User(text);
        var assistant = ChatMessage.Assistant();
        Messages.Add(user);
        Messages.Add(assistant);
        IsResponding = true;

        bool started = current.Send(text, evt => dispatcherQueue.TryEnqueue(() => Apply(evt, assistant)));
        if (!started)
        {
            // Can only happen if a previous turn's background thread has not finished
            // draining yet; IsResponding normally prevents this from being reachable.
            IsResponding = false;
            Messages.Remove(assistant);
            Messages.Remove(user);
            Draft = text;
            ErrorMessage = "The assistant is still finishing the last message.";
        }
    }

    /// <summary>Opens (or reuses) the session for <paramref name="vaultRoot"/>. Fails, with a
    /// message set, when there is no API key yet.</summary>
    private AgentClient? ClientFor(string vaultRoot)
    {
        if (client is { } existing && clientVaultRoot == vaultRoot)
        {
            return existing;
        }

        (string? key, CredentialStoreStatus status) = CredentialStore.TryGetApiKey();
        if (status == CredentialStoreStatus.Unavailable)
        {
            ErrorMessage = "The API key could not be read from Windows Credential Manager.";
            return null;
        }

        if (string.IsNullOrEmpty(key))
        {
            ErrorMessage = "Add an Anthropic API key in Settings to use the assistant.";
            return null;
        }

        client?.Dispose();

        AgentClient? opened = AgentClient.Open(vaultRoot, key);
        if (opened is null)
        {
            ErrorMessage = "Could not open the assistant for this folder.";
            return null;
        }

        client = opened;
        clientVaultRoot = vaultRoot;
        return opened;
    }

    /// <summary>Applies one decoded event to <see cref="Messages"/>. Called on the UI thread.</summary>
    private void Apply(AgentEvent evt, ChatMessage assistant)
    {
        switch (evt.Kind)
        {
            case AgentEventKind.Text:
                assistant.AppendAssistantDelta(evt.Text!);
                break;

            case AgentEventKind.Thinking:
                // Summarised reasoning; not surfaced in this first pass.
                break;

            case AgentEventKind.ToolStarted:
                Messages.Add(ChatMessage.ToolStarted(evt.ToolName!));
                break;

            case AgentEventKind.ToolFinished:
                // Matched from the end: tools can run more than once in a turn, so this finds
                // the most recent still-running row for that name rather than the first.
                ChatMessage? running = Messages
                    .Reverse()
                    .FirstOrDefault(m => m.Kind == ChatMessageKind.Tool && m.ToolName == evt.ToolName && m.ToolOk is null);
                running?.ApplyToolFinished(evt.ToolOk!.Value, evt.Detail!, evt.Commit);
                break;

            case AgentEventKind.Refused:
                IsResponding = false;
                string reason = string.Join(": ", new[] { evt.Category, evt.Explanation }.Where(s => !string.IsNullOrEmpty(s)));
                Messages.Add(ChatMessage.Notice(reason.Length == 0 ? "The assistant declined to continue." : reason));
                break;

            case AgentEventKind.Failed:
                IsResponding = false;
                Messages.Add(ChatMessage.Notice(evt.Message!));
                break;

            case AgentEventKind.Done:
                IsResponding = false;
                break;
        }
    }

    public void Dispose() => client?.Dispose();
}
